### ISSUE TITLE
Line search failing frequently

### ISSUE THREAD

**Issue by KeAWang** [hjmshi/PyTorch-LBFGS#6]
In my use-case of full batch L-BFGS, I'm often running into issues where the line search is is failing. But at the same time I don't want to make the max_ls too large or else the optimizer spends a lot of time trying to find a tiny step size that satisfies the Wolfe conditions, which ends up not affecting the loss function much at all and wasting computation time.

I've tried playing around with the Wolfe condition constants c1 and c2 but it hasn't helped. 

Is there some way to force the optimizer to still take a step at the end of its line-search even if it can't find a step-size that satisfies the Wolfe conditions?

**Comment by hjmshi** [comment]
Thanks for your feedback! We can definitely add the option. Alternatively, you could also try switching to using the FullBatchLBFGS optimizer with the no line search option.

Just to help debug the issue, do you mind giving more details regarding your use case of the method? Is there any error (numerical approximations, stochasticity) introduced in the function values or gradient evaluations? If you increase the max_ls, does the line search eventually find a step size that satisfies the Armijo and Wolfe conditions? Is the line search failing early in the optimization or after optimizing for a while?

Thanks!

**Comment by KeAWang** [comment]
I'm drawing stochastic multivariate Gaussian samples to do a Monte Carlo estimate of a cost function in a RL setting. 

The issue is that optimizers Adam work well but I would like to reduce the number of steps taken by using L-BFGS. I've noticed that increasing the number of samples helps L-BFGS fail less frequently. Unfortunately even then it would fail around 25-50% of the time throughout the optimization (even with max_ls set to 30). Furthermore even when L-BFGS succeeds in its line search, it barely optimizes the loss function at all. For example, Adam is able to get to the desired optimum (after running for many iterations) which has a loss of around 12. On the other hand, regardless of how many iterations it is run for, L-BFGS is consistently unable to decrease the loss below 26 even though the maximum loss possible is 30.

**Comment by hjmshi** [comment]
I see. Adam out of the box is a stochastic optimizer, so it very naturally handles stochastic estimates of the gradient. Unfortunately, full batch L-BFGS out of the box does not directly apply to pure stochastic problems because the quadratic model it builds depends on good estimates of the gradient differences.

Instead, you should be looking to use a stochastic L-BFGS method, which can be done using this module. Do you have any way of controlling which samples you are using when performing your Monte Carlo estimate to ensure consistent sampling between different function and gradient evaluations?

One of the keys to being able to utilize L-BFGS in the stochastic setting is consistent sampling; that is, you must make sure that the functions and gradients used in the L-BFGS approximation and line search are evaluated on the same sample for that particular iteration. If you use different samples for estimating the cost function (and perhaps the gradient), then the line search will be subsequently making decisions incorrectly based off of noise (similarly the L-BFGS approximation will be built on noise).

The best reference to look at for this is the full-overlap or multi-batch examples for neural networks. Please let me know if this is clear.

**Comment by KeAWang** [comment]
I see. I'm now fixing the seed for the closure function so that each iteration of L-BFGS uses the same samples with different samples drawn between iterations. But now line search always succeeds after one closure evaluation even with initial learning rates as large as 100. Also the loss isn't actually decreasing, even though Adam does fine. I followed the example in the full-overlap example exactly. Do you think this might just be a failure of L-BFGS?

**Comment by hjmshi** [comment]
Yes, it could very likely be - I don't have much experience with using L-BFGS in the RL setting. How much variance is there in the function values and gradients? Have you tried using a larger number of samples? Do you see any rapid decrease in the loss at the beginning before reaching some neighborhood? Have you also tried tuning the steplength? 

In the stochastic setting, the line search cannot be totally relied on for selecting the steplength because it doesn't incorporate the stochasticity of the problem - it only guarantees progress on that particular sample and that the L-BFGS model will be well-defined. If you look at some of our prior work (see Bollapragada, et al. (2018)), we actually control/reduce the variance by increasing the batch size via an adaptive mechanism and also use a conservative estimate on the steplength (also dependent on the variance) to get good results. 

Hence, it may be useful to initially try a larger number of samples and use a steplength of 1 or 1/2 (1 is the natural steplength in the deterministic setting for quasi-Newton methods, so using an initial steplength larger than that is not advised). Then (similar to SG methods) increase the number of samples or decrease the steplength once the optimization appears to stagnate (assuming you can make some progress using the first suggestion).

Generally speaking, these stochastic L-BFGS methods work well when the problem is not totally dominated by noise, but only encounters small to moderate amounts of noise. If your problem is not like this, then it's likely that L-BFGS will fail.

By the way, would you mind sharing your problem so we can investigate it further? Our group has been focused on trying to extend quasi-Newton methods to the stochastic setting, so it would be good to see what problems are out there. Thanks!
